### PR TITLE
Onboarding: Library and UI

### LINKED ISSUE #3950: Onboarding feature
### 📒 Description
Onboarding is a feature to present hints in many places (Timer, Timeline, TimeEntry List, Record Activity, ...) if the user matches some conditions, which is described in https://toggl.invisionapp.com/d/main#/console/19683071/411618499/preview

This ticket would contain all onboarding sub-tickets.

### Approach
- Implement a core logic in the Library (count and keep track activity numbers and determine which condition is fulfilled). Thus, we can reuse it later on other platforms.
- Add some extra columns in User table to keep track the number and the current hint state
- Implement a generic Hint layout on macOS
- Present/dismiss logic and highlight the area for the hint 

### 💻 Environment
Platform: macOS & Library

### Tickets 
- [x] [Onboarding] Hint Popover #3951
- [x] [Onboarding] Core logic library #3952
- [x] [Onboarding] Grey out / Highlight Hint area #3953

### LINKED ISSUE #3951: [Onboarding] Hint Popover
### 📒 Description
This ticket would implement the Popover layout for the Hint.

### 💻 Environment
Platform: macOS

### ⭐️ Acceptance Criteria
- Generic Hint layout Popover
- Auto-size depend on the text inside

### LINKED ISSUE #3952: [Onboarding] Core logic library
### 📒 Description
This ticket would implement the core logic on the Library to maintain and keep track of the state and notify the UI to present the hint.
### 💻 Environment
Platform: Library

### ⭐️ Acceptance Criteria
- Add additional columns into User Table (Last time open app, Number of entries, first-time use Timeline, first-time use Recorder, ...). The requirement is described in https://toggl.invisionapp.com/d/main#/console/19683071/411618499/preview
- Expose some helper methods 
- The logic that determine if which hint should be presented

### LINKED ISSUE #3953: [Onboarding] Grey out / Highlight Hint area
### 📒 Description
This ticket would implement the grey out / highlight area when a Hint Popover is presented 

### 💻 Environment
Platform: macOS

### ⭐️ Acceptance Criteria
- Add grey out / highlight area properly, which relies on what is the type of hint

### PR DESCRIPTION
**Pull request by NghiaTranUIT**
### 📒 Description
- This PR will introduce all onboarding logic on the library and improve UI Onboarding on mac
- Design hint: https://toggl.invisionapp.com/d/main#/console/19683071/411618499/inspect
- Don't implement Manual Onboarding Design (As we don't have a Meatball button)

### Summary
- All Onboarding states are store on `onboarding_states` table: Have proper primary key and constraints
- We keep track number of users actions, such as TimelineTab open count, Number of Time Entry, Switch On Record Activity, ..... Determine and trigger the onboarding if there is any requirement fulfilled. See `onboarding_service.cpp` for detail
- Then dispatching the callback to the UI for rendering properly

### 🕶️ Types of changes
**Breaking change** (fix or feature that would cause existing functionality to change

### 🤯 List of changes
- ~Add new columns and migration to User table~ -> Introduce onboarding_states table to store for each user -> To store the onboarding state
- Onboarding service as a singleton to easily access around the codebase
- Reuse EventQueue for the Timer
- The logic for all Onboarding (except the Manual Onboarding)
- Sync the state to the database
- Trigger the UI onboarding
- UI Onboarding on the mac app

#### 1. Couldn't make an corner Rect from BezierPath on CAShapeLayer 
[init(roundedRect:xRadius:yRadius:)](https://developer.apple.com/documentation/appkit/nsbezierpath/1520638-init) causes the corner rect is skew
Screen Shot 2020-04-21 at 14 34 32

### Not done yet
- Corner Rect in the highlight onboarding area

### 👫 Relationships
Closes #3952 
Closes #3950
Closes #3951
Closes #3953 (with the known-issue)

#### 1. Review Onboarding UI on the mac app
1. Add
```
    _callback(OnboardingTypeNewUser);
    return;
```
to the first line of this func https://github.com/toggl-open-source/toggldesktop/blob/63089bb7366a38b73078417878d847b39f0528f8/src/onboarding_service.cpp#L84
in order to manually trigger the UI for testing

2. Start the app
3. Verify that:
- The onboarding is shown properly
- Able to click outside to hide the Popover
- The popover direction matches with the design (The distance between the highlighted area and the Popover's arrow doesn't' match the design since it's super difficult to set the position)
- Resize the window will update the position
- Correct Onboarding text
- We decided to keep the original design of the NSPopover (don't check the background color and the arrow size)
4. Stop the app
5. Change the OnboardingType to different one. Repeat all
Available OnboardingTypes

```
enum OnboardingType {
    OnboardingTypeNewUser,
    OnboardingTypeOldUser,
    OnboardingTypeManualyMode,
    OnboardingTypeTimelineTab,
    OnboardingTypeEditTimeEntry,
    OnboardingTypeTimelineTimeEntry,
    OnboardingTypeTimelineView,
    OnboardingTypeTimelineActivity,
    OnboardingTypeRecordActivity
};
```

##### 2.1: Carefully review the logic
- Review `onboarding_state.cpp` and the Design hint => Make sure the logic is correct

##### 2.1: New User, Timeline View, Timeline Activity Section and Time Entry Onboarding
1. Sign up with new account
2. After 5 seconds from the main app. Verify:
- New User onboarding is shown
2. Click on Timeline Tab. Verify:
- Timeline View onboarding is shown
3. Turn On Record Activity. Verify
- Timeline Activity Section onboarding is shown
4. Start new Time Entry
5. Stop after a couple of minutes. Verify:
- Time Entry onboarding is shown
6. Open the local database (`~/Library/Application Support/Kopsik/development/kopsik.db`) and verify some states is updated
7. Logout the account and log in again
8. Make sure those previous onboarding doesn't show anymore
9. Verify that:
- Old user onboarding is shown

##### 2.2: New/Old user onboarding at the second time
1. Sign up a new account
2. Don't start any TimeEntry
3. Stop the app
4. Open the local database, find the onboarding_state row, which matches with the user_id -> Update the created_at to last week
5. Open the app. Verify
- After 5 seconds, the Old user will present at the second time

##### 2.3 Timeline TimeEntry section onboarding
- Requirement is tricky
```
Displayed if:

> User has never been using recording functionality

and

> They already tracked TEs for 3 days
(so they’ve learned some areas of the app already)

and

> User is on Timeline tab for at least 3rd time
```
1. Sign up with a new account
2. Start any TE and stop on the Timer
3. Switch to Timeline Tab more than 3 times
4. Switch Back to Time Entry List
5. Change the day of the first TE to 4 days ago
6. Stop the app
7. Open the app and click on the Timeline Tab and verify:
- The Onboarding for Timeline Record Activity is shown

##### 2.4 Timeline Time Entry section Onboarding
```
Should be displayed if:

> User visits Timeline tab for the 3rd time

and

> Hasn’t been editing or adding entries directly on the Timeline yet

and

> There is at least one TE to display on chosen day

If there is no TEs to display, hint should be shown on next possible occassion when all the conditions are true
```
1. Sign up new account
2. Start and Stop 1 Time Entry on the Timer
3. (Click on Timeline Tab and Click on TimeEntry List Tab) 3 times
4. Verify:
- At the 3 times on Timeline Tab, the Timeline Time Entry section Onboarding is shown

### CODE REVIEW

**Comment by MartinBriza** [comment]
Do we really need to add more threads to the library? We already have multiple mostly dormant threads for things like window changes or tracking reminders (and probably some more). I'd suggest merging these into one thread with a shared timer so we avoid as much of the threading headache we have right now... 

You can see a possible implementation of something like that here: https://github.com/MartinBriza/toggldesktop/blob/rewrite-internals/next/eventqueue.h
It's a simple timed event queue that runs on a single separate thread and can be safely activated from the outside.

**Comment by NghiaTranUIT** [comment]
I agree that the Timer introduces one more threads, but the Timer function is only executed twice:
1. When the app launches
2. When the condition of Onboarding New User is fulfilled

Thus, I'm not sure that it's reasonable to implement an Event Queue from your suggestion 🤔    

On the other hand, can you help me to improve the current Timer to "close" the thread after the execution? Or reuse your EventQueue is the only choice?

**Comment by MartinBriza** [comment]
My `EventQueue` class is not the only possible implementation of such mechanic, however it's definitely less demanding than yours for these reasons:
 - It starts a new thread only once and keeps it sleeping while there's no activity in the queue - compared to your `Timer` that starts a new thread each time there's an event (and creating a thread is quite resource-intensive)
 - It can be shared between classes
 - the `Stop` method doesn't really stop anything until the timeout passes (and even then, if the callback has already been called and takes a long time, doesn't do anything again)
 - You're using the `Timer` in a way that leaks memory

**Comment by MartinBriza** [comment]
Another thing worth noting (and I invite @IndrekV and @skel35 to voice their opinions) is that adding 14 columns to the `users` table in the database seems a tad excessive. Wouldn't it be better to create a new table just to keep onboarding stuff separate?

**Comment by skel35** [comment]
@MartinBriza I agree, creating a new table looks much more reasonable to me as well.

**Comment by NghiaTranUIT** [comment]
Thank for the initial review 👍 I will put those onboarding state to the new table 👍

**Comment by NghiaTranUIT** [comment]
Ready to review 👍

**Comment by NghiaTranUIT** [comment]
**OnboardingTypeNewUser**
> With the first popup if you type in a description and click on start button it doesn't start. The popup hides with the first click and only second click works to start the timer. First click on the start button should work.

I improve the behavior to pass the click event to the below views, but only if the click position is in the highlighted area. Otherwise, if we click on the grey area, it will close the Popover

> With the second popup that tells the user to click on an entry to start to edit you need to click twice to trigger edit. The first click will close the popup and the second will open edit popup. It should happen with the first click.

As the above change, but I won't work when clicking on the Time Entry CollectionView for some reason. I attempt with `mouseDown:` and `hitTest:`, but there is no solution.

For now, We have to click on to dismiss the Popover and click to select the view.

> I created 3 time entries and put them on 3 different past dates, went to timeline view and did not see the "Record Activity" onboarding. I had been to timeline tab more than 3 times.

Fixed. Every time we change the Timeline Tab, it will receive the start time of the last TE and present the Onboarding if possible.

> Turning "Record Activity" on showed the onboarding the indicated where the data will be visible.

I'm unable to reproduce it. The Popover will trigger when I switch ON the Record Activity

Screen Shot 2020-04-23 at 16 09 44

> Switching to manual mode the first time did not trigger onboarding.

From the Onboarding design, we don't have onboarding when switching to Manual Mode. I suppose that you mean (Screenshot)

Screen Shot 2020-04-23 at 16 10 27

but we decided to drop it since we don't have the Meatball button

**Comment by NghiaTranUIT** [comment]
Please review on the second round @IndrekV 🙌

**Comment by IndrekV** [comment]
The click issues are fixed.

The popup in Timeline view that I'm not seeing is this

Screenshot 2020-04-24 at 12 10 13

**Comment by NghiaTranUIT** [comment]
@IndrekV If this Activity Recording onboarding doesn't present, it means that some conditions are not meet.

https://github.com/toggl-open-source/toggldesktop/blob/b524de9daea9ae78e836985274c14b1e2579ddc5/src/onboarding_service.cpp#L234-L259

It will be triggered when the user selects the Timeline Tab.

I just tested and it works.

### Scenario
1. Sign up with a new account
2. Start new TE -> Stop it
3. Change the TE's date to 3 days ago
4. Click on Timeline Tab and back to List Tab 3 times
5. At the 3 times on Timeline Tab, the onboarding will present

Screen Shot 2020-04-24 at 16 20 25

**Comment by IndrekV** [comment]
Thank you for the description. I'm able to see this popup now.

**Comment by IndrekV** [comment]
The functionality seems ok. I'm seeing crashed from `Eventqueue` when I quit the app I'll try to get more info about it.

**Comment by NghiaTranUIT** [comment]
I'm able to reproduce the crash on Eventqueue @IndrekV. Look like the Destructors of Eventqueue is problematic

Screen Shot 2020-04-24 at 20 33 29


```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff6928033a libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00000001012739bc libsystem_pthread.dylib`pthread_kill + 430
    frame #2: 0x00007fff69207808 libsystem_c.dylib`abort + 120
    frame #3: 0x00007fff66467458 libc++abi.dylib`abort_message + 231
    frame #4: 0x00007fff664587e9 libc++abi.dylib`demangling_terminate_handler() + 48
    frame #5: 0x00007fff66466887 libc++abi.dylib`std::__terminate(void (*)()) + 8
    frame #6: 0x00007fff66466838 libc++abi.dylib`std::terminate() + 56
    frame #7: 0x00007fff664435cd libc++.1.dylib`std::__1::thread::~thread() + 17
  * frame #8: 0x0000000100c4afb6 TogglDesktopLibrary.dylib`toggl::EventQueue::~EventQueue(this=0x000000010706c620) at eventqueue.h:19:7
    frame #9: 0x0000000100c17e75 TogglDesktopLibrary.dylib`toggl::EventQueue::~EventQueue(this=0x000000010706c620) at eventqueue.h:19:7
    frame #10: 0x0000000100c1838e TogglDesktopLibrary.dylib`toggl::Context::~Context(this=0x000000010706c600) at context.cc:176:1
    frame #11: 0x0000000100c18bf5 TogglDesktopLibrary.dylib`toggl::Context::~Context(this=0x000000010706c600) at context.cc:133:21
    frame #12: 0x0000000100c18c1c TogglDesktopLibrary.dylib`toggl::Context::~Context(this=0x000000010706c600) at context.cc:133:21
    frame #13: 0x0000000100d058b0 TogglDesktopLibrary.dylib`::toggl_context_clear(context=0x000000010706c600) at toggl_api.cc:54:5
    frame #14: 0x000000010002b0f3 TogglDesktop`-[AppDelegate applicationWillTerminate:](self=0x0000600003b082a0, _cmd="applicationWillTerminate:", app=@"NSApplicationWillTerminateNotification") at AppDelegate.m:1125:2
    frame #15: 0x00007fff2f1e79cf CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
    frame #16: 0x00007fff2f1e7963 CoreFoundation`___CFXRegistrationPost1_block_invoke + 63
    frame #17: 0x00007fff2f1e78d8 CoreFoundation`_CFXRegistrationPost1 + 372
    frame #18: 0x00007fff2f1e7544 CoreFoundation`___CFXNotificationPost_block_invoke + 80
    frame #19: 0x00007fff2f1b76bd CoreFoundation`-[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1554
    frame #20: 0x00007fff2f1b6b69 CoreFoundation`_CFXNotificationPost + 1351
    frame #21: 0x00007fff3182d866 Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 59
    frame #22: 0x00007fff2c73cc38 AppKit`-[NSApplication terminate:] + 1385
    frame #23: 0x00007fff2c6b1437 AppKit`-[NSApplication(NSResponder) sendAction:to:from:] + 299
    frame #24: 0x00007fff2c7c1aee AppKit`-[NSMenuItem _corePerformAction] + 312
    frame #25: 0x00007fff2c7c186a AppKit`-[NSCarbonMenuImpl performActionWithHighlightingForItemAtIndex:] + 106
    frame #26: 0x00007fff2c7c0767 AppKit`-[NSMenu performKeyEquivalent:] + 412
    frame #27: 0x00007fff2cbf627b AppKit`routeKeyEquivalent + 521
    frame #28: 0x00007fff2c6124dd AppKit`-[NSApplication(NSEvent) sendEvent:] + 1161
    frame #29: 0x00007fff2c45f21f AppKit`-[NSApplication run] + 707
    frame #30: 0x00007fff2c430ff6 AppKit`NSApplicationMain + 777
    frame #31: 0x00000001000208e2 TogglDesktop`main(argc=3, argv=0x00007ffeefbff498) at main.m:13:9
    frame #32: 0x00007fff69138cc9 libdyld.dylib`start + 1
    frame #33: 0x00007fff69138cc9 libdyld.dylib`start + 1
```

**Comment by NghiaTranUIT** [comment]
Look like the patched EventQueue fixes the crash during logout the app since I'm not experiencing the crash anymore 👍 

Ready to review the C++ codebase 🙌

File: src/onboarding_service.cpp [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+#include "database.h"
+#include "user.h"
+#include "time_entry.h"
+#include "related_data.h"
+
+#include <time.h>
+
+namespace toggl {
+
+void OnboardingService::RegisterEventQueue(EventQueue &_queue) {
```
[the remark this answers is not in the document]
Comment (NghiaTranUIT) [reply src/onboarding_service.cpp]: - Removed the Kopsik files
- My purpose in using mutex is that I would like to protect the. `OnboardingState state` from modifying the `state`'s properties across multiple threads. 

I could introduce the internal lock for `OnboardingState` and override the getter/setter to make sure it's safe to access as I did in Swift, but in C++, it's overkill to me.

I'm not sure if the current implementation on a mutex is correct or not. I appreciate the suggestion to fix it in code if it's incorrect.

**Comment by NghiaTranUIT** [comment]
## Update
- Removed EventQueue since I couldn't find the proper solution to fix it (This [attempt](https://github.com/toggl-open-source/toggldesktop/pull/4050) doesn't work well)
- Use lightweight Timer class from https://github.com/eglimi/cpptime
- Replace with a new one and the crash is fixed

It's ready to review 👍

File: src/onboarding_service.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MartinBriza) [reply src/onboarding_service.cpp]: Sice Indrek has already approved this, I'm closing my review.

**Pull request merged**